training/coach.py

```diff
@@ -250,18 +309,76 @@ def get_normed_embeds(self, text_prompts: List[str]) -> torch.Tensor:
                 tokens=text_tokens['input_ids'].long().to(device=self.cfg.device),
                 mask=text_tokens['attention_mask'].to(device=self.cfg.device),
             )
-            text_embs_normed = text_embs / text_embs.norm(p=2, dim=-1, keepdim=True)
+            text_embs_normed = self.normalize_embeds(text_embs)
         return text_embs_normed
 
+    def collect_negative(self, template: list, *, save_image: str, image_embs: Optional[DefaultDict[str, List[int]]], negative_classes: Optional[DefaultDict[str, List[int]]]) -> List[str]:
+        temp = []
+        sampled_image = self.save_images(save_dir=self.cfg.images_root,
+                                        save_prefix=f'{self.train_step}_{save_image}' if save_image else "",
+                                        image_embs=temp,  template=template)
+
+        negatives = self.query_vlm(sampled_image)
+
+        if not image_embs or not negative_classes:
+            return negatives
+
+        classes_temp = []
+        for live_negative in negatives:
+            if not self.cfg.specific_negatives or any( cluster in live_negative for cluster in self.cfg.specific_negatives):
+                classes_temp.append(live_negative)
+            elif image_embs is not None:
+                del temp[len(classes_temp)]
+
+        if temp:
+            image_embs[template[0]].extend(temp)
+            negative_classes[template[0]].extend(classes_temp)
+
+        return negatives
+
+    def get_all_template_embedded(self, save_image: str, image_embs: DefaultDict[str, List[int]], negative_classes: DefaultDict[str, List[int]], count=0):
+        """
+        use all template fomr ConceptDataset
+        which are under training.templates
+        """
+        from training.templates import PREFIXES
+        for temp in ConceptDataset(
+            placeholder_token=self.cfg.placeholder_token,
+            learnable_property=self.cfg.learnable_property
+        ).templates:
+            templates = [temp] if '{a}' not in temp else []
+            if '{a}' in temp:
+                for a in PREFIXES: # a the an
+                    templates.append(temp.format(a=a, token='{token}'))
+            for template in templates:
+                while len(image_embs[template]) <= count:
+                    self.collect_negative([template], save_image=save_image,
+                                                image_embs=image_embs, negative_classes=negative_classes)
+
```

Comment:
get all possible negative image features via the "`templates`"

Reply:
collect image for negative prompt first

1. we will generate image from "mypet" or "pet" prompt
2. put them image feature into image_embs for corresponding class like "a photo of a {token}" but not "a photo of a pet" or "a photo of a dog"